receiver/m365receiver/logs.go

```diff
@@ -0,0 +1,537 @@
+	// create default client
+	httpClient, err := l.cfg.ToClient(host, l.settings)
+	if err != nil {
+		l.logger.Error("error creating HTTP client", zap.Error(err))
+		return err
+	}
+
+	// create m365 log client, create token and start audit subscriptions
+	l.client = newM365Client(httpClient, l.cfg, "https://manage.office.com/.default")
+	err = l.client.GetToken()
+	if err != nil {
+		l.logger.Error("error creating authorization token", zap.Error(err))
+		return err
+	}
+	for _, a := range l.audits {
+		err = l.client.StartSubscription(l.startRoot + a.route)
+		if err != nil {
+			l.logger.Error("error starting audit subscriptions", zap.Error(err))
+			return err
+		}
+	}
+
+	// set cancel function
+	cancelCtx, cancel := context.WithCancel(ctx)
+	l.cancel = cancel
+
+	// init checkpoint stuff
+	storageClient, err := adapter.GetStorageClient(ctx, host, l.cfg.StorageID, l.id)
+	if err != nil {
+		return fmt.Errorf("failed to get storage client: %w", err)
+	}
+	l.storageClient = storageClient
+	l.loadCheckpoint(cancelCtx)
+
+	return l.startPolling(cancelCtx)
+}
+
+func (l *m365LogsReceiver) Shutdown(ctx context.Context) error {
+	l.logger.Debug("shutting down logs receiver")
+	l.cancel()
+	l.wg.Wait()
+	return l.checkpoint(ctx)
+}
+
+// spins a go routine at each poll interval to go get logs
+func (l *m365LogsReceiver) startPolling(ctx context.Context) error {
+	t := time.NewTicker(l.pollInterval)
+	l.wg.Add(1)
+	go func() {
+		defer l.wg.Done()
+		for {
+			select {
+			case <-t.C:
+				if err := l.pollLogs(ctx); err != nil {
+					l.logger.Error("error while polling for logs", zap.Error(err))
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// spins a go routine for each audit type/endpoint
+func (l *m365LogsReceiver) pollLogs(ctx context.Context) error {
+	var st string
+	if l.record.NextStartTime != nil {
+		st = l.record.NextStartTime.Format(layout)
+	} else {
+		st = pcommon.NewTimestampFromTime(time.Now().Add(-l.pollInterval)).AsTime().UTC().Format(layout)
+	}
+
+	now := time.Now().UTC()
+
+	auditWG := &sync.WaitGroup{}
+	auditWG.Add(5)
+	for i := 0; i < len(l.audits); i++ {
+		go l.poll(ctx, now, st, &l.audits[i], auditWG)
+	}
+	auditWG.Wait()
+
+	l.record.NextStartTime = &now
+	return l.checkpoint(ctx)
+}
+
+// collects log data from endpoint, transforms logs, consumes logs
+func (l *m365LogsReceiver) poll(ctx context.Context, now time.Time, st string, audit *auditMetaData, wg *sync.WaitGroup) {
+	defer wg.Done()
+	if !audit.enabled {
+		return
+	}
+
+	l.mu.Lock()
+	data, err := l.client.GetJSON(ctx, l.root+audit.route, now.Format(layout), st)
+	if err != nil {
+		if err.Error() == "authorization denied" { // troubleshoot stale token
+			l.logger.Debug("possible stale token; attempting to regenerate")
+			err = l.client.GetToken()
+			if err != nil { // something went wrong generating token
+				l.logger.Error("error creating authorization token", zap.Error(err))
+				l.mu.Unlock()
```

Comment:
more common pattern is to `defer l.mu.Unlock()` right when you call .Lock(), though in the current form it will hold the lock through the transform & consume code. Tiny refactor to put GetJSON & error handling into a separate function would fix that and be cleaner code.